// mcache/lib/cluster_cache.js
var cluster = require('cluster');
var MemoryCache = require('./lib/memory_cache.js');

function ClusterCache(ttl, options, set_function) {
	this.set_function = set_function;
	this.ttl = ttl;
	this.uid = 
	thi
	
};
ClusterCache.prototype = Object.create(MemoryCache.prototype);

if (cluster.isMaster) {
	
}

function ClusterCache(ttl, options, set_function) {
	this.set_function = set_function;
	this.ttl = ttl;
	this.data = {};
	if (options.gc_time) this.gc_timer = setInterval(ClusterCache.prototype.clearOld.bind(this), options.gc_time);
};

ClusterCache.prototype.get = function(key, callback) {
	if (Array.isArray(key)) {
		this.get_many(key, callback);
	} else {
		this.get_one(key, callback);
	}
};

ClusterCache.prototype.get_many = function(keys, callback) {
	var time = Date.now();
	var to_get = [];
	for(var i=0; i < keys.length; i++) {
		var key = keys[i];
		var obj = this.data[key];
		if (!obj || obj.t && time - obj.t > this.ttl) to_get.push(key);
	}
	if (to_get.length == 0) {
		var result = {};
		for(var i=0; i < keys.length; i++) {
			result[keys[i]] = this.data[keys[i]].v;
		}
		if (callback) callback(null, result);
		return result;
	} else 
		var _this = this;{
		this.set_function(to_get, function(err, data) {
			if (err) {
				if (callback) callback(err);
			} else {
				time = Date.now();
				for(var i=0; i < to_get.length; i++) {
					_this.data[to_get[i]] = {
						v: data[i],
						t: time
					};
				}
				var result = {};
				for(var i=0; i < keys.length; i++) {
					result[keys[i]] = _this.data[keys[i]].v;
				}
				if (callback) callback(null, result);
			}
		});
	}
	return null;
};

ClusterCache.prototype.get_one = function(key, callback) {
	var time = Date.now();
	var obj = this.data[key];
	// obj.t - time when data were setted
	// obj.v - value
	// obj.l - query of get callbacks
	if (obj) {
		if (obj.t) {
			// data is available
			if (time - obj.t > this.ttl) {
				// expired
				delete this.data[key];
				obj = null;
				this.get(key, callback);
			} else {
				if (callback) callback(null, obj.v, obj.t);
				return obj.v;
			}
		} else {
			// no data in memory, but set_function in progress - push callback in query
			if (callback) obj.l.push(callback);
		}
	} else {
		// no data in memory, but set_function in progress - push callback in query
		obj = this.data[key] = {l: []};
		if (callback) obj.l.push(callback);
		var _this = this;
		this.set_function(key, function(err, data) {
			if (err) {
				for(var i in obj.l) {
					obj.l[i](err);
				}
				delete _this.data[key];
				obj = null;
			} else {
				obj.v = data;
				obj.t = Date.now();
				for(var i in obj.l) {
					obj.l[i](null, obj.v, obj.t);
				}
				delete obj.l;
				obj = null;
			}
		});
	}
	return null;
};

ClusterCache.prototype.set = function(key, val) {
	var time = Date.now();
	if (this.data[key] && this.data[key].l) {
		for(var i in this.data[key].l) {
			this.data[key].l[i](null, val, time);
		}
	}
	delete this.data[key];
	this.data[key] = {
		v: val,
		t: time
	};
};

ClusterCache.prototype.del = function(key) {
	delete this.data[key];
};

// update set time
ClusterCache.prototype.updateTime = function(key) {
	this.data[key].t = Date.now();
};

// get keys count
ClusterCache.prototype.size = function() {
	return Object.keys(this.data).length;
};

// get all data in hash
ClusterCache.prototype.hash = function() {
	var r = {};
	for (var i in this.data) {
		if (this.data[i].v) r[i] = this.data[i].v;
	}
	return r;
};

// delete all expired data
ClusterCache.prototype.clearOld = function() {
	var time = Date.now();
	for (var i in this.data) {
		if (this.data[i].t && time - this.data[i].t > this.ttl) {
			delete this.data[i];
		}
	}
};

module.exports = ClusterCache;
